slack-lint/src/main/java/slack/lint/text/SpanPointMarkDangerousCheckDetector.kt

```diff
@@ -0,0 +1,130 @@
+      briefDescription = "Check Span Flags Using Bitmask",
+      explanation = """
+        Spans flags can have priority or other bits set. \
+        Ensure that Span flags are checked using \
+        `currentFlag and Spanned.SPAN_POINT_MARK_MASK == desiredFlag` \
+        rather than just `currentFlag == desiredFlag`
+      """,
+      category = Category.CORRECTNESS,
+      priority = 4,
+      severity = Severity.ERROR,
+      implementation = sourceImplementation<SpanPointMarkDangerousCheckDetector>()
+    )
+  }
+
+  override fun getApplicableUastTypes() = listOf(UBinaryExpression::class.java)
+
+  override fun createUastHandler(context: JavaContext): UElementHandler {
+    return ReportingHandler(context)
+  }
+}
+
+/**
+ * Reports violations of SpanPointMarkDangerousCheck.
+ */
+private class ReportingHandler(private val context: JavaContext) : UElementHandler() {
+  companion object {
+    private const val SPANNED_CLASS = "android.text.Spanned"
+    private val MARK_POINT_FIELDS = setOf(
+      "$SPANNED_CLASS.SPAN_INCLUSIVE_INCLUSIVE",
+      "$SPANNED_CLASS.SPAN_INCLUSIVE_EXCLUSIVE",
+      "$SPANNED_CLASS.SPAN_EXCLUSIVE_INCLUSIVE",
+      "$SPANNED_CLASS.SPAN_EXCLUSIVE_EXCLUSIVE",
+    )
+    private const val MASK_CLASS = "$SPANNED_CLASS.SPAN_POINT_MARK_MASK"
+  }
+
+  override fun visitBinaryExpression(node: UBinaryExpression) {
+    if (node.operator == UastBinaryOperator.EQUALS ||
+      node.operator == UastBinaryOperator.NOT_EQUALS ||
+      node.operator == UastBinaryOperator.IDENTITY_EQUALS ||
+      node.operator == UastBinaryOperator.IDENTITY_NOT_EQUALS
+    ) {
+      checkExpressions(node, node.leftOperand, node.rightOperand)
+      checkExpressions(node, node.rightOperand, node.leftOperand)
+    }
+  }
+
+  private fun checkExpressions(node: UBinaryExpression, markPointCheck: UExpression, maskCheck: UExpression) {
+    if (matchesMarkPoint(markPointCheck) && !matchesMask(maskCheck)) {
+      context.report(
+        ISSUE,
+        context.getLocation(node),
+        """
+          Do not check against ${markPointCheck.sourcePsi?.text} directly. \
+          Instead mask flag with Spanned.SPAN_POINT_MARK_MASK to only check MARK_POINT flags.
+        """.trimIndent(),
+        LintFix.create()
+          .replace()
+          .name("Use bitwise mask")
+          .text(maskCheck.sourcePsi?.text)
+          .with("((${maskCheck.sourcePsi?.text}) and $MASK_CLASS)")
+          .build()
+      )
+    }
+  }
+
+  private fun matchesMarkPoint(expression: UExpression): Boolean {
+    return getQualifiedName(expression) in MARK_POINT_FIELDS
+  }
+
+  private fun matchesMask(expression: UExpression): Boolean {
+    return if (expression is UBinaryExpression) {
+      expression.leftOperand.resolveQualifiedNameOrNull() == MASK_CLASS ||
+        expression.rightOperand.resolveQualifiedNameOrNull() == MASK_CLASS
+    } else {
+      false
+    }
+  }
+
+  private fun getQualifiedName(expression: UExpression): String? {
```

Comment:
Should this be an extension?

Reply:
Do you prefer an extension in util file or private extension in the file?

Reply:
👍 Fixed in 7804c86dbeb96b6c2eb77a0f7e35a9d351619649